common/include/pcl/common/impl/eigen.hpp

```diff
@@ -353,130 +364,30 @@ pcl::eigen33 (const Matrix& mat, Matrix& evecs, Vector& evals)
 
-    tmp = scaledMat;
-    tmp.diagonal ().array () -= evals (1);
-
-    vec1 = tmp.row (0).cross (tmp.row (1));
-    vec2 = tmp.row (0).cross (tmp.row (2));
-    vec3 = tmp.row (1).cross (tmp.row (2));
-
-    len1 = vec1.squaredNorm ();
-    len2 = vec2.squaredNorm ();
-    len3 = vec3.squaredNorm ();
-    if (len1 >= len2 && len1 >= len3)
-    {
-      mmax[1] = len1;
-      evecs.col (1) = vec1 / std::sqrt (len1);
-      min_el = len1 <= mmax[min_el] ? 1 : min_el;
-      max_el = len1 > mmax[max_el] ? 1 : max_el;
-    }
-    else if (len2 >= len1 && len2 >= len3)
-    {
-      mmax[1] = len2;
-      evecs.col (1) = vec2 / std::sqrt (len2);
-      min_el = len2 <= mmax[min_el] ? 1 : min_el;
-      max_el = len2 > mmax[max_el] ? 1 : max_el;
-    }
-    else
+    for (int i = 0; i < 3; ++i)
     {
-      mmax[1] = len3;
-      evecs.col (1) = vec3 / std::sqrt (len3);
-      min_el = len3 <= mmax[min_el] ? 1 : min_el;
-      max_el = len3 > mmax[max_el] ? 1 : max_el;
+      Matrix tmp = scaledMat;
+      tmp.diagonal ().array () -= evals (i);
+      const auto vec_len = detail::getLargest3x3Eigenvector<Vector> (tmp);
+      evecs.col (i) = vec_len.vector;
+      eigenVecLen[i] = vec_len.length;
     }
 
-    tmp = scaledMat;
-    tmp.diagonal ().array () -= evals (0);
-
-    vec1 = tmp.row (0).cross (tmp.row (1));
-    vec2 = tmp.row (0).cross (tmp.row (2));
-    vec3 = tmp.row (1).cross (tmp.row (2));
-
-    len1 = vec1.squaredNorm ();
-    len2 = vec2.squaredNorm ();
-    len3 = vec3.squaredNorm ();
-    if (len1 >= len2 && len1 >= len3)
-    {
-      mmax[0] = len1;
-      evecs.col (0) = vec1 / std::sqrt (len1);
-      min_el = len3 <= mmax[min_el] ? 0 : min_el;
-      max_el = len3 > mmax[max_el] ? 0 : max_el;
-    }
-    else if (len2 >= len1 && len2 >= len3)
-    {
-      mmax[0] = len2;
-      evecs.col (0) = vec2 / std::sqrt (len2);
-      min_el = len3 <= mmax[min_el] ? 0 : min_el;
-      max_el = len3 > mmax[max_el] ? 0 : max_el;
-    }
-    else
-    {
-      mmax[0] = len3;
-      evecs.col (0) = vec3 / std::sqrt (len3);
-      min_el = len3 <= mmax[min_el] ? 0 : min_el;
-      max_el = len3 > mmax[max_el] ? 0 : max_el;
-    }
+    const auto minmax_it = std::minmax_element (eigenVecLen.cbegin (), eigenVecLen.cend ());
+    int min_idx = std::distance (eigenVecLen.cbegin (), minmax_it.first);
+    int max_idx = std::distance (eigenVecLen.cbegin (), minmax_it.second);
+    int mid_idx = 3 - min_idx - max_idx;
 
-    unsigned mid_el = 3 - min_el - max_el;
-    evecs.col (min_el) = evecs.col ( (min_el + 1) % 3).cross (evecs.col ( (min_el + 2) % 3)).normalized ();
-    evecs.col (mid_el) = evecs.col ( (mid_el + 1) % 3).cross (evecs.col ( (mid_el + 2) % 3)).normalized ();
-#ifdef _WIN32
-    delete [] mmax;
-#endif
+    evecs.col (min_idx) = evecs.col ( (min_idx + 1) % 3).cross (evecs.col ( (min_idx + 2) % 3)).normalized ();
+    evecs.col (mid_idx) = evecs.col ( (mid_idx + 1) % 3).cross (evecs.col ( (mid_idx + 2) % 3)).normalized ();
```

Comment:
From what I can tell, by the time you reach this section of the code the eigenvalues under variable `eval` are already sorted from `computeRoots()` and all your eigenvectors are normalized already and in the right columns, from lower to higher eigenvalues.  At this stage, you might or not flip the the sign of 1 eigenvector, to ensure the matrix of all eigenvector has determinant 1 (it is a rotation matrix). You only need to do this for a single eigenvector.

This whole process of deciding which eigenvector to correct based on their length before normalizing, feels like a numerical safeguard.

If my memory does not fail me `pcl::eigen33` is only meant to be used in positive semidefinite matrices (like covariance matrices). Symmetric matrices (a subset of PSD matrices), have orthogonal eigenvectors. In theory, only two eigen vectors need to be explicitly computed. We can leave this for another PR, but I get the feeling that most of this section is redundant.

Reply:
I'll leave it to you to create the issue with the details since I don't know the use-case in details. I did add the same in code to ensure this conversation is not lost

Reply:
@Sergio: I implemented those 2x2 and 3x3 optimizations for eigen-decomposition and inversion. The computeRoots function does assume that the matrix is symmetric as can be seen by the coefficient-calculation of the characteristic equation. Also the eigenvalues are sorted.
This direct and fast method has some short-comings when the matrix is not well-conditioned and should be used with a grain of salt. Similar to the single-look mean&covariance calculation it is ment to be a fast rather than an accurate/robust method.

-Suat

Reply:
@SergioRAgostinho (complete github username is needed for mentioning, else I think the notification goes to the other party)